numpy/core/defchararray.py

```diff
@@ -319,9 +322,18 @@ def add(x1, x2):
     arr1 = numpy.asarray(x1)
     arr2 = numpy.asarray(x2)
     out_size = _get_num_chars(arr1) + _get_num_chars(arr2)
-    dtype = _use_unicode(arr1, arr2)
-    return _vec_string(arr1, (dtype, out_size), '__add__', (arr2,))
 
+    if type(arr1.dtype) != type(arr2.dtype):
+        # Enforce this for now.  The solution to it will be implement add
+        # as a ufunc.  It never worked right on Python 3: bytes + unicode gave
+        # nonsense unicode + bytes errored, and unicode + object used the
+        # object dtype itemsize as num chars (worked on short strings)
+        raise TypeError(
+            "np.char.add() requires both arrays of the same dtype kind, but "
+            f"got dtypes: '{arr.dtype}' and '{arr2.dtype}' (the few cases "
+            "where this used to work often lead to incorrect results).")
```

Comment:
After some back and forth, I decided that trying to get the logic right seems the wrong approach.  So I nuked it.  The things that didn't already fail were:

* bytes + unicode.  Which converted the unicode to bytes (all 4 for each character) and then stored it into a too short unicode result as a byte string.
* unicode + object (where the object array contains strings).  Which would always use `np.dtype("O").itemsize` for the number of characters, and thus shorten longer object strings anways.

Neither seems reasonable to me.

Reply:
bytes+void made a little bit more sense, but even in that case you strip NUL padding of the void by casting it to bytes.  The promotion of void to bytes doesn't really make sense...